README.md
#Memcached [![Build Status](https://secure.travis-ci.org/3rd-Eden/node-memcached.png?branch=master)](http://travis-ci.org/3rd-Eden/node-memcached)

`memcached` is a fully featured Memcached client for Node.js. `memcached` is
build with scaling, high availability and exceptional performance in mind. We
use consistent hashing to store the data across different nodes. Consistent
hashing is a scheme that provides a hash table functionality in a way that
adding or removing a server node does not significantly change the mapping of
the keys to server nodes. The algorithm that is used for consistent hashing is
the same as `libketama`.

There are different ways to handle errors for example, when a server becomes
unavailable you can configure the client to see all requests to that server as
cache misses until it goes up again. It's also possible to automatically remove
the affected server from the consistent hashing algorithm or provide `memcached`
with a failover server that can take the place of the unresponsive server.

When these issues occur the `memcached` client will emit different events where
you can subscribe to containing detailed information about the issues.

The client is configurable on different levels. There's a global configuration
that you update so all you Memcached clusters will use the same failure
configuration for example, but it's also possible to overwrite these changes per
`memcached` instance.

### protocol

This module uses the ASCII protocol to communicate with the server, this makes
it easier to debug for you are user as you can see what is send over the wire
but also for me as developer. But this also means that SASL auth is not
supported in this driver as that requires the use of the binary protocol. The
ASCII protocol not only used by memcached but also by other databases and
message queues, so that is a nice extra.

## Setting up the client

The constructor of the `memcached` client take 2 different arguments `server
locations` and `options`. Syntax:

``` js
var Memcached = require('memcached');
var memcached = new Memcached(Server locations, options);
```

### Server locations

The server locations is designed to work with different formats. These formats
are all internally parsed to the correct format so our consistent hashing scheme
can work with it. You can either use:

1. **String**, this only works if you have are running a single server instance
   of Memcached.  It's as easy a suppling a string in the following format:
   `hostname:port`. For example `192.168.0.102:11212` This would tell the client
   to connect to host `192.168.0.102` on port number `11212`.

2. **Array**, if you are running a single server you would only have to supply
  one item in the array.  The array format is particularly useful if you are
  running a cluster of Memcached servers. This will allow you to spread the keys
  and load between the different servers. Giving you higher availability for
  when one of your Memcached servers goes down.
  
3. **Object**, when you are running a cluster of Memcached servers it could
   happen to not all server can allocate the same amount of memory. You might
   have a Memcached server with 128mb, 512, 128mb. If you would the array
   structure all servers would have the same weight in the consistent hashing
   scheme. Spreading the keys 33/33/33 over the servers. But as server 2 has
   more memory available you might want to give it more weight so more keys get
   stored on that server. When you are using a object, the `key` should
   represent the server location syntax and the value the weight of the server.
   By default all servers have a weight of 1.  `{ '192.168.0.102:11212': 1,
   '192.168.0.103:11212': 2, '192.168.0.104:11212': 1 }` would generate a
   25/50/25 distribution of the keys.

If you would implement one of the above formats, your constructor would
something like this:

```js
var memcached = new Memcached({ '192.168.0.102:11212': 1, '192.168.0.103:11212': 2, '192.168.0.104:11212': 1 });
var memcached = new Memcached([ '192.168.0.102:11212', '192.168.0.103:11212', '192.168.0.104:11212' ]);
var memcached = new Memcached('192.168.0.102:11212');
```

### Options

There 2 kinds of options that can be configured. A global configuration that
will be inherited by all Memcached servers instances and a client specific
configuration that can be used to overwrite the globals. The options should be
formatted in an JavaScript `object`. They both use the same object structure:

* `maxKeySize`: *250*, the max size of they key allowed by the Memcached server.
* `maxExpiration`: *2592000*, the max expiration of keys by the Memcached server
  in milliseconds.
* `maxValue`: *1048576*, the max size of a value that is allowed by the
  Memcached server.
* `poolSize`: *10*, the maximum connections we can allocate in our connection pool.
* `algorithm`: *crc32*, the hashing algorithm that should be used to generate
  the hashRing values.
* `reconnect`: *18000000*, when the server is marked as dead we will attempt to
  reconnect every x milliseconds.
* `timeout`: *5000*, after x ms the server should send a timeout if we can't
  connect. This will also be used close the connection if we are idle.
* `retries`: *5*, amount of tries before we mark the server as dead.
* `retry`: *30000*, timeout between each retry in x milliseconds.
* `remove`: *false*, when the server is marked as dead you can remove it from
  the pool so all other will receive the keys instead.
* `failOverServers`: *undefined*, the ability use these servers as failover when
  the dead server get's removed from the consistent hashing scheme. This must be
  an array of servers confirm the server_locations specification.
* `keyCompression`: *true*, compress keys using md5 if they exceed the
  maxKeySize option.

Example usage:

```js
var memcached = new Memcached('localhost:11212', {retries:10,retry:10000,remove:true,failOverServers:['192.168.0.103:11212']});
```

If you wish to configure the options globally:

```js
var Memcached = require('memcached');
// all global configurations should be applied to the .config object of the Client.
Memcached.config.poolSize = 25;
```

## API

### Public methods


### Private methods
The following methods are intended for private usage:

---------------------------------------
#### .connect
Fetches or generates a connection for the given server. The supplied callback
function will receive a reference to the connection as argument.
If there are issues with the server connection, we are going to respond with cache-miss pattern.

**Arguments**

`server`: *String*, The server that needs a connection, the format must be
confirm the server_locations specification.

`callback`: *Function*, The callback function that receives the net.Stream
connection. It will be called with 2 arguments `error` and `connection`.

Example:

``` js
memcached.connect( '192.168.0.103:11212', function( err, conn ){
  if( err ) throw new Error( err );
  console.log( conn.server );
});
```

---------------------------------------
#### .multi
A small wrapper function that makes it easier to query multiple Memcached
servers. It will return the location for each key or the complete list of
servers.

**Arguments**

`keys`: *Array* **(optional)**, They keys that needs to be converted to a server.

`callback`: *Function*, The callback function for the data, it will be called
for **each** key. It will be called with 4 arguments:

1. `server`: *String*, The server location.
2. `key`: *String*, The key associated with the server, if you didn't specify
   keys, this variable will be undefined.
3. `index`: *Number*, The current index of the loop
4. `total`: *Number*, The total amount server retrieved.

Example:

``` js
memcached.multi( false, function( server, key, index, totals ){
  if( err ) throw new Error( err );
  
  this.connect( server, function( err, conn ){
    console.log( "connection ready" )
  })
});
```

---------------------------------------
#### .command

This is the core functionality of the `memcached` client. All public API's are
routed through this function. It takes care of the argument validations Server
retrieval ( If the server argument isn't specified ). After all data ready a
connection is asked for the private `connect` method and the command is written
to the Memcached server.

**Arguments**

`query`: *Object*, The metaData object, see the `Callbacks` section for the
specification.

`server`: *String*, The server the to connect. This is only needed when the
metaData object doesn't contain a key property to retrieve the server from. 

Example:

``` js
memcached.command({
  key: 'key', callback: function(){ console.dir( arguments ); },

  // validate the arguments
  validate: [[ 'key', String ], [ 'callback', Function ]],

  // used for the query
  type: 'delete',
  command: 'delete key'
});
```

---------------------------------------
#### .connectionIssue

A internal function for logging issues with connections. As there can be various
of ways that an error occurs we need solid issue manager to handle all these
cases. For example server could crash or the Memcached server could respond with
`SERVER ERROR <broken>`.

**Arguments**

`error`: *String*, The actual error message.

`Stream`: *net.Stream*, A reference to the connection stream where the error
occurred on.

`callback`: *Function* **(optional)**, The callback function of a potential
request, it will be marked as cache miss if it was provided

Example:

``` js
memcached.connectionIssue( "Server down", connectionReference );
```

## Callbacks

Each method requires a callback function. Once this function get executed there
will be 2 variables applied:

* `error`: A error response if something went wrong while retrieving data from
  the Memcached server. Depending on the type of request this will either be an
  string or an Array with multiple errors.
* `response`: The actual result from the Memcached server. If the response is
  `false` or `undefined` than a cache miss occurred. Cache misses will also
  occur when there is an error. So you might want to check on errors first.

When we have a successful response, the context of the callback function will
shift to a metaData object. The metaData object contains all information that we
used to generate the request for the Memcached server. The metaData object
contains the following properties:

* `start`: Date in milliseconds when the request was received 
* `execution`: Total execution time for the request, including response parsing.
* `callback`: Reference to the callback function
* `type`: The type of Memcached command
* `command`: The compiled command that was send through the sockets
* `validate`: The properties of metaData object that needs type validation.

And all the arguments you have send to the method, this depends on the method
you have called. 
  
## Events

When connection issues occur we send out different notifications using the
`EventEmitter` protocol. This can be useful for logging, notification and
debugging purposes. Each event will receive details Object containing detailed
information about the issues that occurred. 

### Details Object

The details Object contains the various of error messages that caused, the
following 3 will always be present in all error events:

* `server`: the server where the issue occurred on
* `tokens`: a array of the parsed server string in `[port, hostname]` format.
* `messages`: a array containing all error messages that this server received.
  As messages are added to the array using .push(), the first issue will at the
  beginning and the latest error at the end of the array.

The following properties depend on the type of event that is send. If we are
still in our retry phase the details will also contain:

* `retries`: the amount of retries left before we mark the server as dead.
* `totalRetries`: the total amount of retries we did on this server, as when the
  server has been reconnected after it's dead the `retries` will be rest to
  defaults and messages will be removed.

If the server is dead these details will be added:

* `totalReconnectsAttempted`: the total reconnects we have attempted. This is
the success and failure combined.
* `totalReconnectsSuccess`: the total successful reconnects we have made.
* `totalReconnectsFailed`: the total failed reconnects we have made.
* `totalDownTime`: the total down time that was generated. Formula: (
  totalReconnectsFailed * reconnect_timeout ) + ( totalRetries * retry_timeout).

### Events

There are `5` different events that the `memcached` client emits when connection
issues occur. 

* `issue`: a issue occurred on one a server, we are going to attempt a retry next.
* `failure`: a server has been marked as failure or dead.
* `reconnecting`: we are going to attempt to reconnect the to the failed server.
* `reconnected`: successfully reconnected to the memcached server.
* `remove`: removing the server from our consistent hashing.

Example implementations:

```js
var memcached = new Memcached([ '192.168.0.102:11212', '192.168.0.103:11212' ]);
memcached.on('failure', function( details ){ sys.error( "Server " + details.server + "went down due to: " + details.messages.join( '' ) ) });
memcached.on('reconnecting', function( details ){ sys.debug( "Total downtime caused by server " + details.server + " :" + details.totalDownTime + "ms")});
```


CHANGELOG
0.2.1
 - Supports for a queued callback limit so it would crash the process when we queue
   to much callbacks. #81

0.2.0
 - [breaking] We are now returning Error instances instead of strings for errors
 - Dependency bump for a critical bug in our connection pool.

0.1.5
 - Don't execute callbacks multiple times if the connection fails
 - Parser fix for handling server responses that contain Memcached Procotol
   keywords
 - Make sure that the retry option is set correctly

0.1.4
 - Added missing error listener to the 3rd-Eden/jackpot module, this prevents crashes
   when it's unable to connect to a server.

0.1.3
 - Handle Memcached responses that contain no value.
 - Travis CI integration.

0.1.2
 - Returning an error when the Memcached server issues a NOT_STORED response.

0.1.1
 - Now using 3rd-Eden/jackpot as connection pool, this should give a more stable
   connection.

0.1.0
 - Storing numeric values are now returned as numeric values, they are no
   longer strings.

0.0.12
 - Added .setEncoding to the connections, this way UTF-8 chars will not be
   chopped in to multiple pieces and breaking binary stored data or UTF-8 text

0.0.11
 - Added more useful error messages instead of returning false. Please note
   that they are still strings instead of Error instances (legacy)

0.0.10
 - Compatibility with Node.js 0.8
 - Don't saturate the Node process by retrying to connect if pool is full #43
 - Minor code formatting

0.0.9
 - Codestyle refactor, named the functions, removed tabs
 - Added Mocha test suite


client.js
var EventEmitter = require('events').EventEmitter,
	Stream		 = require('net').Stream,
	Buffer		 = require('buffer').Buffer,
	CreateHash	 = require('crypto').createHash;

var HashRing 	 = require('./lib/hashring').HashRing,
	Connections  = require('./lib/connection'),
	Utils		 = require('./lib/utis'),
	Manager		 = Connection.Manager,
	IssueLog	 = Connection.IssueLog;

exports.Client = function Client( args, options ){
	var servers = [],
		weights = {},
		key;
	
	// parse down the connection arguments	
	switch( Object.prototype.toString.call( args ) ){
		case '[object String]':
			servers.push( args );
			break;
		case '[object Object]':
			weights = args;
			servers = Object.keys( args );
		case '[object Array]': 
		default:
			servers = args;
			break;
	}
	
	// merge with global and user config
	Utils.merge( this, Client.config );
	Utils.merge( this, options );
	EventEmitter.call( this );
	
	this.servers = servers;
	this.HashRing = new HashRing( servers, weights );
	this.connections = [];
	this.issues = [];
};

// Allows users to configure the memcached globally or per memcached client
Client.config = {
	max_key_size: 250,			 // max keysize allowed by Memcached
	max_expiration: 2592000,	 // max expiration duration allowed by Memcached
	max_value: 1048576,			 // max length of value allowed by Memcached
	
	pool_size: 10,				 // maximal parallel connections
	reconnect: 18000000,		 // if dead, attempt reconnect each xx ms
	retries: 5,					 // amount of retries before server is dead
	retry: 30000,				 // timeout between retries, all call will be marked as cache miss
	remove: false,				 // remove server if dead if false, we will attempt to reconnect
	
	compression_threshold: 10240,// only than will compression be usefull
	key_compression: true		 // compress keys if they are to large (md5)
};

// There some functions we don't want users to touch so we scope them
(function( nMemcached ){
	const LINEBREAK				= '\r\n',
		  FLAG_JSON 			= 1<<1,
		  FLAG_COMPRESSION 		= 2<<1,
		  FLAG_JCOMPRESSION 	= 3<<1,
		  FLAG_BINARY 			= 4<<1,
		  FLAG_COMPRESSEDBINARY = 5<<1,
		  FLAG_ESCAPED			= 6<<1;
	
	var memcached = nMemcached.prototype = new EventEmitter;
	
	memcached.connect = function( server, callback ){
		if( server in this.issues && this.issues[ server ].failed )
			return callback( false, false );
			
		if( server in this.connections )
			return this.connections[ server ].allocate( callback );
		
		var server_tokens = /(.*):(\d+){1,}$/.exec( server ).reverse(),
			memcached = this;
			server_tokens.pop();
		
		this.connections[ server ] = new Manager( server, this.pool_size, function( callback ){
			var S = new Stream,
				Manager = this;
			
			// config the Stream
			S.setTimeout(0);
			S.setNoDelay(true);
			S.metaData = [];
			S.server = server;
			S.tokens = server_tokens;
			
			Utils.fuse( S, {
				connect	: function(){ callback( false, this ) },
				close	: function(){ Manager.remove( this ) },
				error	: function( err ){ memcached.connectionIssue( error, S, callback ) },
				data	: memcached.rawDataReceived,
				end		: S.end
			});
			
			// connect the net.Stream [ port, hostname ]
			S.connect.apply( S, server_tokens );
		});
		
		this.connections[ server ].allocate( callback );
	};
	
	memcached.command = function( query ){
		var server = this.HashRing.get_node( query.key );
		
		if( server in this.issues && this.issues[ server ].failed )
			return callback( false, false );
		
		this.connect( server, function( error, S ){
			if( !S ) return query.callback( false, false );
			if( error ) return query.callback( error );
			if( S.readyState !== 'open' ) return query.callback( 'Connection readyState is set to ' + S.readySate );
			
			S.metaData.push( query );
			S.write( query.command + LINEBREAK );
		})
	};
	
	memcached.connectionIssue = function( error, S, callback ){
		// end connection and mark callback as cache miss
		S.end(); callback( false, false );
		
		var issues,
			server = S.server,
			memcached = this;
		
		// check for existing issue logs, or create a new	
		if( server in this.issues ){
			issues = this.issues[ server ];
		} else {
			issues = this.issues[ server ] = new IssueLog({
				server: 	server,
				tokens: 	S.tokens,
				reconnect: 	this.reconnect,
				retries: 	this.retries,
				retry: 		this.retry,
				remove: 	this.remove
			});
			
			// proxy the events
			Utils.fuse( issues, {
				issue:			function( details ){ memcached.emit( 'issue', details ) },
				failure: 		function( details ){ memcached.emit( 'failure', details ) },
				reconnecting: 	function( details ){ memcached.emit( 'reconnecting', details ) },
				reconnected: 	function( details ){ memcached.emit( 'reconnect', details ) },
				remove: 		function( details ){
									// emit event and remove servers
									memcached.emit( 'remove', details );
									memcached.connections[ server ].end();
									
									if( this.failOverServers && this.failOverServers.length )
										memcached.HashRing.replaceServer( server, this.failOverServers.shift() );
									else
										memcached.HashRing.removeServer( server );
								}
			});
		}
		
		// log the issue
		issues.log( error );
	};
	
	memcached.disconnect = function(){
		this.connections.forEach(function( Manager ){ Manager.free(0) });
	};
	
	
	
})( Client )